mlflow/gateway/providers/palm.py

```diff
@@ -0,0 +1,189 @@
+from fastapi.encoders import jsonable_encoder
+
+from mlflow.gateway.config import PaLMConfig, RouteConfig
+from mlflow.gateway.providers.base import BaseProvider
+from mlflow.gateway.providers.utils import rename_payload_keys, send_request
+from mlflow.gateway.schemas import chat, completions, embeddings
+
+
+class PaLMProvider(BaseProvider):
+    def __init__(self, config: RouteConfig) -> None:
+        super().__init__(config)
+        if config.model.config is None or not isinstance(config.model.config, PaLMConfig):
+            raise TypeError(f"Unexpected config type {config.model.config}")
+        self.palm_config: PaLMConfig = config.model.config
+
+    async def _request(self, path: str, payload: Dict[str, Any]) -> Dict[str, Any]:
+        return await send_request(
+            headers={},
+            base_url="https://generativelanguage.googleapis.com/v1beta3/models/",
+            path=path + f"?key={self.palm_config.palm_api_key}",
+            payload=payload,
+        )
+
+    async def chat(self, payload: chat.RequestPayload) -> chat.ResponsePayload:
+        payload = jsonable_encoder(payload, exclude_none=True)
+        self.check_for_model_field(payload)
+        key_mapping = {
+            "stop": "stopSequences",
+            "candidate_count": "candidateCount",
+            "max_tokens": "maxOutputTokens",
+        }
+        for k1, k2 in key_mapping.items():
+            if k2 in payload:
+                raise HTTPException(
+                    status_code=422, detail=f"Invalid parameter {k2}. Use {k1} instead."
+                )
+        payload = rename_payload_keys(payload, key_mapping)
+
+        # Replace 'role' with 'author' in payload
+        for m in payload["messages"]:
+            if "role" in m:
+                m["author"] = m.pop("role")
+
+        # Map 'messages', 'examples, and 'context' to 'prompt'
+        prompt = {"messages": payload.pop("messages")}
+        if "examples" in payload:
+            prompt["examples"] = payload.pop("examples")
+        if "context" in payload:
+            prompt["context"] = payload.pop("context")
+        payload["prompt"] = prompt
+
+        resp = await self._request(
+            f"{self.config.model.name}:generateMessage",
+            payload,
+        )
+
+        # Response example
+        # (https://developers.generativeai.google/api/rest/generativelanguage/models/generateMessage)
+        # ```
+        # {
+        #   "candidates": [
+        #     {
+        #       "author": "1",
+        #       "content": "Hi there! How can I help you today?"
+        #     }
+        #   ],
+        #   "messages": [
+        #     {
+        #       "author": "0",
+        #       "content": "hi"
+        #     }
+        #   ]
+        # }
+        # ```
+        return chat.ResponsePayload(
+            **{
+                "candidates": [
+                    {
+                        "message": {
+                            "role": c["author"],
```

Comment:
Is `author` always present?

Reply:
Yes! If `author` isn't provided, they automatically return `0`, `1`, based on ordering

Reply:
https://developers.generativeai.google/api/rest/generativelanguage/Message

<img width="873" alt="image" src="https://github.com/mlflow/mlflow/assets/17039389/96538c3b-097a-4900-85fa-3030f9d5ddc3">

This doc says it's optinal, but doesn't mention when it's missing :)

Reply:
btw is `0` or `1` acceptable as `role`?

Reply:
It's acceptable because we accept strings. Should we map 1 or to something else (e.g. system)?

Reply:
Yeah was also unsure when it's missing since it seems to be present even when we don't provide the `author` field. Currently, there are no restrictions on what we input/output for `role`. However, for the Databricks AI Gateway, we strictly enforce that `role` must be one of `system`, `user`, `assistant`.